src/raw_rtde/connection.rs

```diff
@@ -0,0 +1,891 @@
+use std::io::{Cursor, ErrorKind, Read, Write};
+use std::mem::{discriminant, Discriminant};
+use std::net::SocketAddr;
+use std::ops::Add;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+use mio::net::TcpStream;
+use mio::{event, Events, Interest, Poll, Token, Waker};
+
+use crate::raw_rtde::packet_data::{ControllerVersion, IoSetup, RTDEData};
+
+const RTDE_SEND_TIMEOUT: Duration = Duration::from_millis(50);
+const RTDE_WAIT_FOR_REPLY_TIMEOUT: Duration = Duration::from_millis(2000);
+
+enum RTDECommand {
+    ConfigureInputs = b'I' as isize,
+    ConfigureOutputs = b'O' as isize,
+    Start = b'S' as isize,
+    Pause = b'P' as isize,
+    RtdeDataPacket = b'U' as isize,
+    RequestProtocolVersion = b'V' as isize,
+    GetControllerVersion = b'v' as isize,
+}
+const WOKEN_BEFORE_READ_PSEUDO_CMD: u8 = 255;
+
+/// A low-level class to control RTDE at the level of single packets.
+///
+/// Except for tests it's recommended to use `RawRTDE` to send RTDE over TCP/IP,
+/// as the UR robot expects to receive it.
+pub struct GenericRawRTDE<T: Read + Write + event::Source> {
+    stream: T,
+    poll: Poll,
+    waker: Arc<Waker>,
+    wait_for_reply_timeout: Duration, // Overridden in tests.
+}
+
+/// A raw RTDE connection over TCP/IP.
+pub type RawRTDE = GenericRawRTDE<TcpStream>;
+
+const RTDE_VERSION: u16 = 2;
+const HEADER_SIZE: usize = 3;
+
+const SOCKET_WAIT_TOKEN: Token = Token(0);
+const WAKE_TOKEN: Token = Token(1);
+
+enum ReadResult {
+    ReadData,
+    WokenBeforeRead,
+}
+
+/// Common method of a real RawRTDE and FakeRawRTDE.
+pub trait RawRTDETrait {
+    /// Informs the robot to start sending data. `configure_outputs` need to be called first.
+    fn start_robot_to_host_transmition(&mut self) -> Result<(), String>;
+    /// Asks the robot to stop sending data.
+    fn pause_robot_to_host_transmition(&mut self) -> Result<(), String>;
+    /// Specifies which RTDE variables we want to write to.
+    ///
+    /// There can be several RTDE connections to the robot, but each variable can be
+    /// written to by at most one connection.
+    fn configure_inputs<U: 'static + AsRef<str>>(
+        &mut self,
+        variables: &[U],
+    ) -> Result<IoSetup, String>;
```

Comment:
`String` is not a great error type. Use `thiserror` for boilerplate-free semantic error types.